File: structure/gcd.java
class Gcd
{

    /**
     * 辗转相除法(欧几里得算法) 递归实现
     * 
     * @param x
     * @param y
     * @return
     */
    public int gcd(int x, int y) {
        if (x < y)
            return gcd(y, x);
        return y == 0 ? x : gcd(y, x % y);
    }

    /**
     * 辗转相除法(欧几里得算法) 循环实现
     * 
     * @param x
     * @param y
     * @return
     */
    public int gcd1(int x, int y) {
        if (x < y)
            return gcd(y, x);
        int temp;
        while (y != 0) {
            temp = x % y;
            x = y;
            y = temp;
        }
        return x;
    }

    /**
     * Stein 算法 循环实现
     * 
     * @param x
     * @param y
     * @return
     */
    public int Stein(int x, int y) {
        int c_2 = 0, temp;
        while ((x & 1) != 1 && (y & 1) != 1) {
            x = x >>> 1;
            y = y >>> 1;
            c_2++;
        }
        while ((x & 1) != 1) x = x >>> 1;
        while ((y & 1) != 1) y = y >>> 1;
        if (x < y) {
            temp = x;
            x = y;
            y = temp;
        }
        while (x == ((x - y) >>> 1)) {
            while ((x & 1) != 1) x = x >>> 1;
            if (x < y) {
                temp = x;
                x = y;
                y = temp;
            }
        }
        return y << c_2;
    }

    /**
     * Stein 算法 递归实现
     * 
     * @param x
     * @param y
     * @return
     */
    public int Stein1(int x, int y) {
        if (y > x)
            return Stein1(y, x);
        if (y == 0)
            return x;
        if ((x & 1) != 1) {
            if ((y & 1) != 1)
                return Stein1(x >> 1, y >> 1) << 1;
            else
                return Stein1(x >> 1, y);
        } else {
            if ((y & 1) != 1)
                return Stein1(x, y >> 1);
            else
                return Stein1((x - y) >> 1, y);
        }
    }
}

File: structure/gcd.cpp
#include <bits/stdc++.h>

#define CHECK(a) (!(1 & (a))) //判断是否被2整除
#define CLEAN2(a)    \
    while (CHECK(a)) \
    a = a >>= 1 //移除非公因子的2
#define BIGERA \
    if (a < b) \
    (t = a, a = b, b = t) //取较大的数为a

/**
 * Stein 算法 循环实现
 */
int gcd(int a, int b)
{
    int c_2 = 0, t;
    while ((CHECK(a)) && (CHECK(b)))
    {
        a = a >>= 1;
        b = b >>= 1;
        c_2++;
    }
    CLEAN2(a);
    CLEAN2(b);
    BIGERA;
    while (a == ((a - b) >> 1))
    {
        CLEAN2(a);
        BIGERA;
    }
    return b << c_2;
}

/**
 * Stein 算法 递归实现
 */
int gcd(unsigned int x, unsigned int y)
{
    if (y > x)
        return gcd(y, x);
    if (!y)
        return x;
    if (!(x & 1))
    {
        if (!(y & 1))
            return gcd(x >> 1, y >> 1) << 1;
        else
            return gcd(x >> 1, y);
    }
    else
    {
        if (!(y & 1))
            return gcd(x, y >> 1);
        else
            return gcd((x - y) >> 1, y);
    }
}

/**
 * 辗转相除法(欧几里得算法)
 */
int gcd(int a, int b)
{
    if (a < b)
        std::swap(a, b);
    return b == 0 ? a : gcd(b, a % b);
}